### PR TITLE
[REFACTOR] fixes #2967 - refactored Fact Importer

### PR DESCRIPTION
**Pull request by ohadlevy**
Based on Dmitri Dolguikh dmitri@appliedlogic.ca original patch.
- renamed original fact_importer class to fact_parser class
- moved fact handling logic to its own fact_importer class
- fact importer now assumes single-value facts
- added a uniqueness constraint on fact_values table on fact_name_id and host_id
- added tests

replaces PR https://github.com/theforeman/foreman/pull/646

### CODE REVIEW

File: app/services/puppet_fact_importer.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+  def initialize(host, facts = {})
+    @host = host
+    @facts = normalize(facts)
+    @counters = {}
+  end
+
+  # expect a facts hash
+  def import!
+
```
Comment (ares) [inline app/services/puppet_fact_importer.rb]: nitpick - empty line

File: app/services/puppet_fact_importer.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    logger.info("Import facts for '#{host}' completed. Added: #{counters[:added]}, Updated: #{counters[:updated]}, Deleted #{counters[:deleted]} facts")
+  end
+
+  private
+  attr_reader :host, :facts
+
+  def delete_removed_facts
+    deleted_counter = FactValue.delete_all([
+                                             'fact_name_id IN (SELECT ID FROM fact_names WHERE name NOT IN (?)) AND host_id=?',
```
Comment (ares) [inline app/services/puppet_fact_importer.rb]: nitpick - id should be downcase I think

File: app/services/puppet_fact_importer.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  def normalize(facts)
+    facts.keep_if { |k, v| v.present? && v.is_a?(String) }
+  end
+
+  def db_facts
+    @db_facts ||= host.fact_values.includes(:fact_name).index_by(&:name)
+  end
+
+  def current_facts_counter
+    FactValue.where(:host_id => host).count
```
Comment (ares) [inline app/services/puppet_fact_importer.rb]: maybe `host.fact_values.count` so you don't have to specify foreign key name

**Comment by ohadlevy** [comment]
@ares fixed all of your comments so far :)

File: app/services/puppet_fact_importer.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  private
+  attr_reader :host, :facts
+
+  def delete_removed_facts
+    deleted_counter     = FactValue.delete_all([
+                                                 'fact_name_id IN (SELECT id FROM fact_names WHERE name NOT IN (?)) AND host_id=?',
+                                                 facts.keys, host.id
+                                               ])
```
Comment (ohadlevy) [inline app/services/puppet_fact_importer.rb:27]: I agree this ignores all callbacks, this would become important if we start auditing facts

**Comment by ares** [comment]
I tested it and it works nicely. So besides `delete_all` I really like this.

**Comment by ohadlevy** [comment]
What would you prefer instead of delete_all just destroy_all ?

**Comment by ares** [comment]
Yes and in such case I would also remove nested SQL query

``` ruby
deleted_counter = host.fact_values.joins(:fact_name).where('fact_names.name NOT IN (?)', facts.keys).destroy_all.size
```

or more readable

``` ruby
to_delete       = host.fact_values.joins(:fact_name).where('fact_names.name NOT IN (?)', facts.keys)
deleted_counter = to_delete.destroy_all.size
```

**Comment by dmitri-d** [comment]
lgtm

**Comment by ohadlevy** [comment]
merged as fb68fd2 with @ares suggestions

**Pull request closed without merging**